com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/StepRequestHandler.java

```diff
@@ -0,0 +1,219 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package com.microsoft.java.debug.core.adapter.handler;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import com.microsoft.java.debug.core.DebugEvent;
+import com.microsoft.java.debug.core.DebugUtility;
+import com.microsoft.java.debug.core.IDebugSession;
+import com.microsoft.java.debug.core.adapter.AdapterUtils;
+import com.microsoft.java.debug.core.adapter.ErrorCode;
+import com.microsoft.java.debug.core.adapter.IDebugAdapterContext;
+import com.microsoft.java.debug.core.adapter.IDebugRequestHandler;
+import com.microsoft.java.debug.core.protocol.Events;
+import com.microsoft.java.debug.core.protocol.Messages.Response;
+import com.microsoft.java.debug.core.protocol.Requests.Arguments;
+import com.microsoft.java.debug.core.protocol.Requests.Command;
+import com.microsoft.java.debug.core.protocol.Requests.StepArguments;
+import com.sun.jdi.Location;
+import com.sun.jdi.Method;
+import com.sun.jdi.StackFrame;
+import com.sun.jdi.ThreadReference;
+import com.sun.jdi.event.StepEvent;
+import com.sun.jdi.request.StepRequest;
+
+public class StepRequestHandler implements IDebugRequestHandler {
+    private Map<Long, ThreadState> threadStates;
+
+    public StepRequestHandler() {
+        threadStates = new HashMap<>();
+    }
+
+    @Override
+    public List<Command> getTargetCommands() {
+        return Arrays.asList(Command.CONFIGURATIONDONE, Command.STEPIN, Command.STEPOUT, Command.NEXT);
+    }
+
+    @Override
+    public CompletableFuture<Response> handle(Command command, Arguments arguments, Response response,
+            IDebugAdapterContext context) {
+        if (context.getDebugSession() == null) {
+            return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.EMPTY_DEBUG_SESSION, "Debug Session doesn't exist.");
+        }
+
+        if (command == Command.CONFIGURATIONDONE) {
+            context.getDebugSession().getEventHub().stepEvents().subscribe(debugEvent -> {
+                handleDebugEvent(debugEvent, context.getDebugSession(), context);
+            });
+        } else {
+            long threadId = ((StepArguments) arguments).threadId;
+            ThreadReference thread = DebugUtility.getThread(context.getDebugSession(), threadId);
+            if (thread != null) {
+                setPendingStepKind(thread, command);
+                setOriginalStackDepth(thread);
+                setOriginalStepLocation(thread);
+                if (command == Command.STEPIN) {
+                    DebugUtility.stepInto(thread, context.getDebugSession().getEventHub(), context.getDebugFilters().stepFilters);
```

Comment:
Why isn't anyone using the returned request object?

Reply:
fixed. the latest change has used the stepRequest info.